vllm/model_executor/layers/quantization/ops/awq.py

```diff
@@ -0,0 +1,281 @@
+) -> torch.Tensor:
+    """Matrix multiplication for AWQ quantized weights.
+
+    Args:
+        a: An input activation tensor of shape `(M, K)`. FP16 or BF16.
+        b: A packed weight tensor of shape `(K, N//P)`. INT32.
+        qzeros: A tensor of shape `(K//G, N//P)`. INT32.
+        scales: A tensor of shape `(K//G, N)`. FP16 or BF16.
+        pack_factor: The packing factor abbreviated as `P`.
+        group_size: The quantization group size abbreviated as `G`.
+        shifter: A tensor of shape `(P,)`. INT32. The shifter for unpacking
+            the packed weight tensor.
+    """
+    if pack_factor != 8:
+        raise NotImplementedError("AWQ pack factor must be 8.")
+    if group_size != 128:
+        raise NotImplementedError("AWQ group size must be 128.")
+    if shifter is None:
+        shifter = get_shifter(pack_factor)
+
+    # Check if the tensors are contiguous.
+    assert a.is_contiguous()
+    assert b.is_contiguous()
+    assert qzeros.is_contiguous()
+    assert scales.is_contiguous()
+
+    # Check dtypes.
+    assert a.dtype in (torch.float16, torch.bfloat16)
+    assert b.dtype == torch.int32
+    assert qzeros.dtype == torch.int32
+    assert scales.dtype == a.dtype
+
+    # Check shapes.
+    assert a.shape[1] == b.shape[0]
+    M, K = a.shape
+    _, PACKED_N = b.shape
+    P = pack_factor
+    N = P * PACKED_N
+    G = group_size
+    assert qzeros.shape == (K // G, PACKED_N)
+    assert scales.shape == (K // G, N)
+
+    # Allocate output.
+    c = torch.empty((M, N), dtype=a.dtype, device=a.device)
+
+    # Launch kernel.
+    dot_out_dtype = tl.float32
+    grid = lambda META: (triton.cdiv(M, META['BLOCK_M']) * triton.cdiv(
+        N, META['BLOCK_N']), META['SPLIT_K'])
+    _awq_kernel[grid](a,
+                      b,
+                      c,
+                      M,
+                      N,
+                      K,
+                      qzeros,
+                      scales,
+                      shifter,
+                      a.stride(0),
+                      a.stride(1),
+                      b.stride(0),
+                      b.stride(1),
+                      c.stride(0),
+                      c.stride(1),
+                      qzeros.stride(0),
+                      qzeros.stride(1),
+                      scales.stride(0),
+                      scales.stride(1),
+                      P,
+                      G,
+                      dot_out_dtype=dot_out_dtype,
+                      GROUP_M=8)
+    return c
+
+
+_SHIFTER_CACHE: Dict[int, torch.Tensor] = {}
+
+
+def get_shifter(pack_factor: int) -> torch.Tensor:
+    assert pack_factor == 8
+    if pack_factor in _SHIFTER_CACHE:
+        return _SHIFTER_CACHE[pack_factor]
+
+    shifter = torch.tensor([0, 4, 1, 5, 2, 6, 3, 7],
+                           dtype=torch.int32,
+                           device="cuda")
+    shifter *= 4
+    _SHIFTER_CACHE[pack_factor] = shifter
+    return shifter
```

Comment:
@WoosukKwon could you help explain this shifting for the Triton kernel? 

- Is the `shifter *= 4` related to the fact that it is 4-bit?
- Can we do `shifter *= 3` if we wanted to implement AWQ with 3-bit precision?
- Is the `[0, 4, 1, 5, 2, 6, 3, 7]` a generic array for shifting and why is it different from the one in AutoAWQ. Is it a simplification of [this loop in AutoAWQ](https://github.com/casper-hansen/AutoAWQ/blob/main/awq/modules/linear.py#L173-180)?